mm2src/coins/eth.rs

```diff
@@ -2775,23 +2791,32 @@ impl EthCoin {
         tx: &[u8],
         swap_contract_address: Address,
         search_from_block: u64,
-    ) -> Result<Option<FoundSwapTxSpend>, String> {
-        let unverified: UnverifiedTransaction = try_s!(rlp::decode(tx));
-        let tx = try_s!(SignedEthTx::new(unverified));
+    ) -> Result<Option<FoundSwapTxSpend>, MmError<FoundSwapTxSpendErr>> {
+        let unverified: UnverifiedTransaction =
+            rlp::decode(tx).map_to_mm(|err| FoundSwapTxSpendErr::Internal(err.to_string()))?;
+        let tx = SignedEthTx::new(unverified).map_to_mm(|err| FoundSwapTxSpendErr::Internal(err.to_string()))?;
 
         let func_name = match self.coin_type {
             EthCoinType::Eth => "ethPayment",
             EthCoinType::Erc20 { .. } => "erc20Payment",
         };
 
-        let payment_func = try_s!(SWAP_CONTRACT.function(func_name));
-        let decoded = try_s!(payment_func.decode_input(&tx.data));
+        let payment_func = SWAP_CONTRACT
+            .function(func_name)
+            .map_to_mm(|err| FoundSwapTxSpendErr::ContractCallError(err.to_string()))?;
```

Comment:
`ContractCallError` is also not a common error type for every coin. Consider implementing `From<Error> for FoundSwapTxSpendError`